lib/vast/Dialect/HighLevel/Transforms/HLEmitLazyRegions.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#include "vast/Dialect/HighLevel/Passes.hpp"
+
+VAST_RELAX_WARNINGS
+#include <mlir/Analysis/DataLayoutAnalysis.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/IR/BlockAndValueMapping.h>
+VAST_UNRELAX_WARNINGS
+
+#include "PassesDetails.hpp"
+#include "vast/Conversion/Common/Passes.hpp"
+#include "vast/Conversion/Common/Patterns.hpp"
+
+#include "vast/Dialect/HighLevel/HighLevelAttributes.hpp"
+#include "vast/Dialect/HighLevel/HighLevelTypes.hpp"
+#include "vast/Dialect/HighLevel/HighLevelOps.hpp"
+
+#include "vast/Dialect/Core/CoreOps.hpp"
+
+#include "vast/Util/TypeList.hpp"
+#include "vast/Util/DialectConversion.hpp"
+
+namespace vast::hl
+{
+    using conversion_rewriter = mlir::ConversionPatternRewriter;
+
+    template< typename source, typename LOp >
+    struct bin_lop_pattern : operation_conversion_pattern< source > {
+        using base = operation_conversion_pattern< source >;
+        using base::base;
+        using adaptor_t = typename source::Adaptor;
+
+        logical_result matchAndRewrite(source op, adaptor_t adaptor, conversion_rewriter &rewriter) const override {
+
+            mlir::BlockAndValueMapping mapper;
+
+            // Extracting the return type from the yield at the end of the block.
+            auto lhs_lazy = rewriter.create< core::LazyOp >(
+                op.getLoc(), op.getLhs().back().back().getOperand(0).getType()
+            );
+            op.getLhs().cloneInto(&lhs_lazy.getLazy(), mapper);
+
+            auto rhs_lazy = rewriter.create< core::LazyOp >(
+                op.getLoc(), op.getRhs().back().back().getOperand(0).getType()
+            );
+            op.getRhs().cloneInto(&rhs_lazy.getLazy(), mapper);
```

Comment:
Copy & paste of the `lhs` case, you can always wrap it into lambda `[ & ]( auto& side ) { ... replace getRhs with side`.